### PR TITLE
Speed up the listing of rule groups on the Ruler Configuration API

### LINKED ISSUE #3121: Ruler: GET /api/v1/rules is slow
We have observed consistent slowness (> 3s) when fetching the list of the rules/alerts from this endpoint. Today, while around the area I noticed that we do a `List` of objects in the bucket followed up by a `Get` of each in serial. In the case I mention, we have about 63 rule groups.

Looking around the current object storage we don't support batching but I think the Thanos one does (?). This might be yet another reason to perhaps move in that direction. If not, it feels like we need support for batching. 

Being fair, I'm not swinging to one solution or another just weighting down our options here.

### PR DESCRIPTION
**Pull request by gotjosh**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #3121 

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pstibrany** [comment]
I'm currently changing how rules are being loaded for shuffle sharding, and will need to only do it for some users, not all users. I'm thinking modifying `List` method to add `delimeter` parameter directly to it, to allow both recursive and non-recursive listings. Different users will then be loaded concurrently.

We may want to coordinate our efforts.

File: pkg/chunk/gcp/gcs_object_client.go
Comment (pstibrany) [inline pkg/chunk/gcp/gcs_object_client.go:115]: This should be:
```suggestion
    q.SetAttrSelection([]string{"Name", "Updated"})
``` 

`SetAttrSelection` method will check and map these names to ones expected by API.

File: pkg/ruler/rules/objectclient/rule_store.go
Comment (pstibrany) [inline pkg/ruler/rules/objectclient/rule_store.go:101]: Given that `ListAllRuleGroups` returns first error, this gouroutine can stop after sending error to the channel.

File: pkg/ruler/rules/objectclient/rule_store.go
Comment (pstibrany) [inline pkg/ruler/rules/objectclient/rule_store.go:81]: Can we use [errgroup](https://godoc.org/golang.org/x/sync/errgroup) here? That should remove need for `WaitGroup` and error channel, and will stop all goroutines when context is cancelled.

(Unbuffered channel here will cause leak of gouroutines if multiple workers try to send error)

File: pkg/ruler/rules/objectclient/rule_store.go
Comment (pstibrany) [inline pkg/ruler/rules/objectclient/rule_store.go:109]:
```suggestion
				mtx.Lock()
				userGroupMap[user] = append(userGroupMap[user], rg)
				mtx.Unlock()
```

1. `defer` executes when function leaves, but this function doesn't until reading channel completely
2. we can simplify adding to the map

File: pkg/ruler/rules/objectclient/rule_store.go
Comment (pstibrany) [inline pkg/ruler/rules/objectclient/rule_store.go:119]: By returning here without closing the channel, worker goroutines may get stuck on reading from channel.

I'd suggest 1) using buffered channel, with buffer equal to `len(ruleGroupObjects)`, 2) pushing all elements to the channel, and closing it, and only then 3) wait until all goroutines finish. Using `errgroup` will make this wait context-aware as well.

**Review by pstibrany — CHANGES_REQUESTED** [review]
Thank you. Please take a look at comments.

File: pkg/chunk/gcp/gcs_object_client.go
Comment (gotjosh) [reply pkg/chunk/gcp/gcs_object_client.go]: Good catch.

**Comment by gotjosh** [comment]
@pstibrany I have a feeling this should wait until #3235 is merged so that I can rebase and pick up the changes.

File: pkg/ruler/rules/objectclient/rule_store.go
Comment (pstibrany) [inline pkg/ruler/rules/objectclient/rule_store.go:170]: You may want to put this debug into fetching goroutine just before calling `o.getRuleGroup(gCtx, key)`, to avoid logging too early.

File: pkg/ruler/rules/objectclient/rule_store.go
Comment (pstibrany) [inline pkg/ruler/rules/objectclient/rule_store.go:215]:
```suggestion
	return result, err
```

File: pkg/ruler/rules/objectclient/rule_store.go
Comment (pstibrany) [inline pkg/ruler/rules/objectclient/rule_store.go:104]: No need for extra `if` here... `nil` is a valid empty slice to return. Using just

```
    return groups[userID], nil`
```

or even
```
    return groups[userID], err
```

and removing previous `if err != nil` will work just fine.

**Review by pstibrany — APPROVED** [review]
Good job, thanks! I've left few nit comments. Integration test failure is unrelated.

File: pkg/ruler/rules/objectclient/rule_store.go
Comment (gotjosh) [reply pkg/ruler/rules/objectclient/rule_store.go:215]: Doing it like this would return partial results on `result`, but I guess it does not really matter given uses of this check for  `err` and return immediately.

File: pkg/ruler/rules/objectclient/rule_store.go
Comment (gotjosh) [reply pkg/ruler/rules/objectclient/rule_store.go:104]: I tested that in a playground with unsuccessful results https://play.golang.org/p/3-iF20WnP2I

Could be it be that Go's type system coerces nil to the empty return type so that `len` works?

Comment (pstibrany) [reply pkg/ruler/rules/objectclient/rule_store.go]: Try this instead: https://play.golang.org/p/BGOaFKkQkzR

`nil` interface{} is different than `nil` slice or `nil` map.

Comment (gotjosh) [reply pkg/ruler/rules/objectclient/rule_store.go]: Thanks! I have addressed the changes.

**Review by pstibrany — APPROVED** [review]
Thank you!

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:43]: It's not a change impacting the user. Also, it's not even a change for embedders. It's more an `[ENHANCEMENT]`. We should also mention it affects only listing objects in the bucket.

```suggestion
* [ENHANCEMENT] Chunks GCS object storage client uses the `fields` selector to limit the payload size when listing objects in the bucket. #3218
```

**Review by pracucci — APPROVED** [review]
LGTM! Just left a nit about the CHANGELOG entry.

**Pull request merged**
